### PR TITLE
[IMP] delivery_state: automatic email sending

### PR DESCRIPTION
**Pull request by chienandalu**
In order to send automatic notifications to the customer when the picking is
confirmed:

  1. Go to *Inventory > Configuration > Settings*.
  2. Enable the option *Send Delivery Confirmation*.

If enabled, an automatic notification will be sent to the picking customer.

cc @Tecnativa TT27723

### CODE REVIEW

File: delivery_state/i18n/es.po [hunk 1]
```diff
@@ -16,6 +16,84 @@ msgstr ""
...
+"        "
+msgstr ""
+"\n"
+"<div>\n"
+"    % set partner = object.sale_id.partner_id or object.partner_id\n"
+"    <p>\n"
+"        Hola, ${partner.name}.\n"
+"    </p>\n"
+"    <p>\n"
+"        Nos complace informarle de que su pedido ha sido enviado por ${object.carrier_id.name}.\n"
```
Comment (pedrobaeza) [inline delivery_state/i18n/es.po:64]:
```suggestion
"        Nos complace informarle que su pedido ha sido enviado por ${object.carrier_id.name}.\n"
```

File: delivery_state/i18n/es.po [hunk 2]
```diff
@@ -16,6 +16,84 @@ msgstr ""
...
+"        % endif\n"
+"    </p>\n"
+"</div>\n"
+"\n"
+"        "
+
+#. module: delivery_state
+#: model:mail.template,subject:delivery_state.delivery_notification
+msgid "${object.company_id.name} - Your order has been shipped by ${object.carrier_id.name}"
+msgstr "${object.company_id.name} - Su pedido ha sido enviado por ${object.carrier_id.name}"
```
Comment (pedrobaeza) [inline delivery_state/i18n/es.po:95]:
```suggestion
msgstr "${object.company_id.name} - Su pedido ha sido enviado a través de ${object.carrier_id.name}"
```

File: delivery_state/i18n/es.po [hunk 3]
```diff
@@ -39,7 +127,7 @@ msgstr "Fecha de entrega"
 #. module: delivery_state
 #: model:ir.model,name:delivery_state.model_delivery_carrier
 msgid "Delivery Methods"
-msgstr "Métodos de entrega"
+msgstr "Método de envío"
```
Comment (pedrobaeza) [inline delivery_state/i18n/es.po:130]:
```suggestion
msgstr "Métodos de envío"
```

File: delivery_state/models/res_company.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+# Copyright 2021 Tecnativa - David Vidal
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import fields, models
+
+
+class ResCompany(models.Model):
+    _inherit = "res.company"
+
+    send_delivery_confirmation = fields.Boolean(
+        string="When the delivery is done send a confirmation email "
```
Comment (pedrobaeza) [inline delivery_state/models/res_company.py:10]:
```suggestion
        string="When the delivery is done, send a confirmation email "
```

File: delivery_state/models/res_config_settings.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# Copyright 2021 Tecnativa - David Vidal
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+from odoo import fields, models
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = "res.config.settings"
+
+    send_delivery_confirmation = fields.Boolean(
+        string="Send Delivery Confirmation",
```
Comment (pedrobaeza) [inline delivery_state/models/res_config_settings.py:10]: Put this as string in the company field, and don't put anything here and the related "magic" will happen

File: delivery_state/tests/test_delivery_state.py [hunk 6]
```diff
@@ -2,56 +2,63 @@
 # Copyright 2020 FactorLibre
```
Comment (pedrobaeza) [inline delivery_state/tests/test_delivery_state.py:2]: Add your copyright in the proper places

File: delivery_state/models/stock_picking.py [hunk 7]
```diff
@@ -73,3 +73,44 @@ def _update_delivery_state(self):
                 lambda x: x.delivery_type == delivery_type)
             for picking in delivery_type_pickings:
                 getattr(picking.carrier_id, method)(picking)
+
+    def _get_delivery_mail_context(self):
+        """Extensible context for customization"""
+        delivery_template_id = self.env.ref(
+            "delivery_state.delivery_notification").id
```
Comment (pedrobaeza) [inline delivery_state/models/stock_picking.py:80]:
```suggestion
            "delivery_state.delivery_notification", raise_if_not_found=False).id
```

File: delivery_state/models/stock_picking.py [hunk 8]
```diff
@@ -73,3 +73,44 @@ def _update_delivery_state(self):
...
+        super().send_to_shipper()
+        partner = self.sale_id.partner_id or self.partner_id
+        send_delivery = all({
+            self.company_id.send_delivery_confirmation,
+            partner,
+            self.carrier_id,
+        })
+        if not send_delivery:
+            return
+        self.force_shipping_confirmation_email_send()
```
Comment (pedrobaeza) [inline delivery_state/models/stock_picking.py:116]:
```suggestion
        if send_delivery:
            self.force_shipping_confirmation_email_send()
```

**Review by pedrobaeza — COMMENTED** [review]
Good job. We have a very complete module for sales shippings. Only some fine-tuning needed.

File: delivery_state/i18n/es.po (hunk 1 above)
Comment (chienandalu) [reply delivery_state/i18n/es.po:64]: https://www.fundeu.es/recomendacion/queismo/

**Comment by chienandalu** [comment]
Changes done

File: delivery_state/i18n/es.po (hunk 1 above)
Comment (pedrobaeza) [reply delivery_state/i18n/es.po:64]: Right, I always fall in the same one:

https://forum.wordreference.com/threads/informar-informarle-de-que.2470670/

That's why foreign students hate Spanish...

**Review by pedrobaeza — APPROVED** [review]

**Review by marcelsavegnago — APPROVED** [review]

**Review by victoralmau — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Please synchronize 13.0 branch with all the missing changes from 12.0

**Pull request merged**